RecoTracker/PixelSeeding/plugins/alpaka/CAHitNtuplet.cc

```diff
@@ -20,57 +22,230 @@
+      auto cellSoA = product.view<::reco::CAGraphSoA>();
+      auto modulesSoA = product.view<::reco::CAModulesSoA>();
+
+      for (int i = 0; i < n_modules; ++i) {
+        auto det = dets[i];
+        auto vv = det->surface().position();
+        auto rr = Rotation(det->surface().rotation());
+        modulesSoA[i].detFrame() = Frame(vv.x(), vv.y(), vv.z(), rr);
+      }
+
+      for (int i = 0; i < n_layers; ++i) {
+        layerSoA.layerStarts()[i] = layerStarts[i];
+        layerSoA.caThetaCut()[i] = iCache->caThetaCuts_[i];
+        layerSoA.caDCACut()[i] = iCache->caDCACuts_[i];
+      }
+
+      layerSoA.layerStarts()[n_layers] = layerStarts[n_layers];
+
+      for (int i = 0; i < n_pairs; ++i) {
+        cellSoA.graph()[i] = {{uint32_t(iCache->pairGraph_[2 * i]), uint32_t(iCache->pairGraph_[2 * i + 1])}};
+        cellSoA.phiCuts()[i] = iCache->phiCuts_[i];
+        cellSoA.minz()[i] = iCache->minZ_[i];
+        cellSoA.maxz()[i] = iCache->maxZ_[i];
+        cellSoA.maxr()[i] = iCache->maxR_[i];
+        cellSoA.startingPair()[i] = false;
+      }
+
+      for (const int& i : iCache->startingPairs_)
+        cellSoA.startingPair()[i] = true;
+
+      return std::make_shared<CAGeometryCache>(std::move(product));
+    }
+
+    static std::unique_ptr<::reco::CAGeometryParams> initializeGlobalCache(edm::ParameterSet const& iConfig) {
+      return std::make_unique<::reco::CAGeometryParams>(iConfig.getParameterSet("geometry"));
+    }
+
   private:
     const edm::ESGetToken<MagneticField, IdealMagneticFieldRecord> tokenField_;
-    const device::ESGetToken<PixelCPEFastParams<TrackerTraits>, PixelCPEFastParamsRecord> cpeToken_;
     const device::EDGetToken<HitsOnDevice> tokenHit_;
     const device::EDPutToken<TkSoADevice> tokenTrack_;
 
+    const TFormula maxNumberOfDoublets_;
+    const TFormula maxNumberOfTuples_;
     Algo deviceAlgo_;
   };
 
   template <typename TrackerTraits>
-  CAHitNtupletAlpaka<TrackerTraits>::CAHitNtupletAlpaka(const edm::ParameterSet& iConfig)
+  CAHitNtupletAlpaka<TrackerTraits>::CAHitNtupletAlpaka(const edm::ParameterSet& iConfig,
+                                                        const ::reco::CAGeometryParams* iCache)
       : EDProducer(iConfig),
         tokenField_(esConsumes()),
-        cpeToken_(esConsumes(edm::ESInputTag("", iConfig.getParameter<std::string>("CPE")))),
         tokenHit_(consumes(iConfig.getParameter<edm::InputTag>("pixelRecHitSrc"))),
         tokenTrack_(produces()),
-        deviceAlgo_(iConfig) {}
+        maxNumberOfDoublets_(
+            TFormula("doubletsHitsDependecy", iConfig.getParameter<std::string>("maxNumberOfDoublets").data())),
+        maxNumberOfTuples_(
+            TFormula("tracksHitsDependency", iConfig.getParameter<std::string>("maxNumberOfTuples").data())),
+        deviceAlgo_(iConfig) {
+    iCache->tokenGeometry_ = esConsumes<edm::Transition::BeginRun>();
+    iCache->tokenTopology_ = esConsumes<edm::Transition::BeginRun>();
```

Comment:
`iCache` is a `const*`, why are changes made to it's pointed-to data ?

Reply:
See also https://github.com/cms-sw/cmssw/issues/34353